### PR TITLE
test(multimodal): harden SSRF rejection paths

### PR DESCRIPTION
**Pull request by pskiran1**
## Overview:

This PR closes SSRF error-propagation gaps for multimodal media URLs. When Dynamo rejects a media URL, the rejection must remain terminal. No adapter, connector, batch loader, or fallback should retry the URL or continue fetching other media.
This PR fixes paths where a typed rejection could be swallowed or treated as recoverable. It also adds request-level tests that use a real network listener to confirm that blocked destinations receive zero connection attempts.

## Details:

- SGLang’s NVDEC adapter path now re-raises URL-validation and HTTP-status errors instead of handling them as generic decode failures and falling back to URL passthrough.
- Rust’s media loader now returns typed `InvalidArgument` errors for policy violations, including blocked schemes, hostnames, IP ranges, DNS-resolved private addresses, and redirect destinations.
- Rust redirect errors preserve policy classification, while ordinary DNS, connection, and transport failures remain non-policy errors.
- Rust multimodal routing now preserves terminal versus recoverable image-dimension failures through its cache. Policy rejections stop the request, while ordinary image-header failures can still use the existing routing fallback.
- Added a shared loopback TCP canary for request-level tests. It counts connection attempts and tests that blocked destinations receive zero outbound connections.
- Added adapter and batch tests for image, audio, and video loaders, the vLLM media connector, SGLang, and TensorRT-LLM.
- The adapter tests inject both `UrlValidationError` and `HttpStatusError`, verify that the same error instance escapes, and confirm that fallback and downstream fetch methods are not called.
- Added a direct comparison between the Python and Rust production IP and hostname blocklists.
- Added a Python backend-decoding request test that confirms a blocked URL returns 4xx with zero outbound connections.
- Added a GPU request-level test that starts a real frontend and worker in both vLLM frontend-processor and Rust frontend-decoding topologies. Both must return 4xx and make zero connections to the blocked destination.

## Where should the reviewer start?
- `lib/llm/src/preprocessor/media/loader.rs` — typed Rust policy errors, redirect handling, DNS checks, and fetcher tests.
- `lib/llm/src/preprocessor.rs` — preservation of terminal policy errors through the dimension cache and MM-routing URL-passthrough path.
- `components/src/dynamo/sglang/request_handlers/multimodal/encode_worker_handler.py` — the terminal-error guard in `_maybe_nvdec_decoder`.
- `components/src/dynamo/common/tests/multimodal/` — batch propagation and vLLM connector fallback contracts.
- `components/src/dynamo/trtllm/tests/test_trtllm_multimodal_processor.py` — exact error propagation and prevention of later media fetches.
- `tests/frontend/test_http_status_propagation.py` — Python backend request-level 4xx and zero-egress coverage.
- `tests/mm_router/test_vllm_frontend_ssrf_e2e.py` — vLLM and Rust frontend request-level coverage.
- `tests/utils/network_canary.py` — shared connection-counting test helper.

## Related Issues

> ⚠️ **This section is required.** Choose one path below and delete the other.

**🔗 This PR is linked to an issue:**

- Closes DIS-2670

**🚫 This PR is NOT linked to an issue**:
- [X] Confirmed — no related issue

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved protection against server-side request forgery (SSRF) by consistently blocking disallowed media URLs, hosts, ports, IP addresses, redirects, and DNS targets.
  * Invalid media URLs now return clear client errors instead of falling back to potentially unsafe URL handling.
  * HTTP fetch failures and URL validation errors now propagate accurately during image, audio, and video processing.
  * Prevented blocked requests from triggering unnecessary media fetching, codec probing, or downstream processing.
* **Tests**
  * Expanded coverage for multimodal media errors, SSRF prevention, and network-egress blocking across supported processing paths.

### CODE REVIEW

**Review by nv-yna — APPROVED** [review]

**Review by nnshah1 — APPROVED** [review]

File: components/src/dynamo/sglang/request_handlers/multimodal/encode_worker_handler.py [hunk 1]
```diff
@@ -644,21 +644,21 @@ async def _maybe_nvdec_decoder(self, url: str) -> Optional[Any]:
             # Constructing the decoder opens the container and reads its frame
             # index, so keep it off the event loop.
             return await asyncio.to_thread(NvdecVideoDecoder, content)
-        except UrlValidationError:
-            # A policy refusal is not a decode failure and must not degrade to
-            # the URL fallback. SGLang applies no policy of its own: it fetches
-            # http(s) through get_mm_http_session and resolves file:// to a bare
-            # path, so passing a rejected URL on turns "denied" into an
-            # unvalidated fetch or local read.
+        except (UrlValidationError, HttpStatusError):
```
Comment (rmccorm4) [inline components/src/dynamo/sglang/request_handlers/multimodal/encode_worker_handler.py]: **[P1] Keep every failed fetch terminal** — Only `UrlValidationError` and `HttpStatusError` are rethrown. Timeout, connection, and other `HttpError` failures reach the broad fallback, return `None`, and cause the caller to pass the original URL to SGLang for an unprotected second fetch. This preserves a DNS-rebinding/redirect SSRF path. Treat every protected-fetch failure as terminal; only decoder failures after bytes are available may fall back.

File: tests/frontend/test_http_status_propagation.py [hunk 2]
```diff
@@ -85,3 +96,46 @@ def test_http_status_propagates_through_wire(services: int) -> None:
...
+                        },
+                    ],
+                }
+            ],
+        },
+        timeout=30,
+    )
+
+    assert 400 <= response.status_code < 500, response.text
+    assert "blocked range" in response.text
```
Comment (rmccorm4) [inline tests/frontend/test_http_status_propagation.py]: **[P1] Exercise the Python rejection path** — The Rust frontend rejects this direct-IP URL before the fake Python worker runs, returning `Direct IP access is not allowed`; therefore this assertion fails CI and does not cover Python backend decoding. Configure a pass-through request topology that reaches `ImageLoader` while retaining the zero-egress canary, then assert the Python-specific error.

File: lib/llm/src/preprocessor.rs [hunk 3]
```diff
@@ -2861,10 +2891,6 @@ impl OpenAIPreprocessor {
             return Ok((w, h));
         }
 
-        if !(url.starts_with("http://") || url.starts_with("https://")) {
-            anyhow::bail!("unsupported url scheme for dim fetch: {}", url);
-        }
-
         // `DIM_FETCH_MEDIA_FETCHER` and `DIM_FETCH_HTTP_CLIENT` are
         // module-scope `LazyLock`s forced at startup in `new_with_parts`
         // for MM-routable preprocessors — see their definitions for the
```
[the remark this answers is not in the document]
Comment (pskiran1) [reply lib/llm/src/preprocessor.rs:3184]: Good catch, this was not intended. I restored the scheme guard ahead of the policy check, so `s3/gs/file` are skipped again instead of failing the request. Updated `url_passthrough_policy_rejection_is_terminal` to use a blocked IP instead of `ftp://`, and added a test covering the passthrough schemes.

File: tests/mm_router/test_vllm_frontend_ssrf_e2e.py [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+
+@pytest.fixture(scope="module", params=["vllm_processor", "rust_decoding"])
+def frontend_topology(
+    request, runtime_services, predownload_models
+) -> Generator[tuple[str, int], None, None]:
+    topology = request.param
+    frontend_port, system_port = allocate_ports(count=2, start_port=13000)
+    with _VllmWorkerProcess(request, topology=topology, system_port=system_port):
+        time.sleep(2)  # let discovery settle before the frontend registers
+        with _FrontendProcess(request, topology=topology, frontend_port=frontend_port):
```
[the remark this answers is not in the document]
Comment (pskiran1) [reply tests/mm_router/test_vllm_frontend_ssrf_e2e.py:166]: Agreed, removed. The frontend health check already waits on `/v1/models`, so the sleep added nothing. I also made that check wait for this specific model id rather than any model.

File: components/src/dynamo/sglang/request_handlers/multimodal/encode_worker_handler.py (hunk 1 above)
Comment (pskiran1) [reply components/src/dynamo/sglang/request_handlers/multimodal/encode_worker_handler.py]: Agreed, thank you. Moved the fetch out of the try so every fetch failure propagates. Only decoder failures after the bytes are in hand fall back, and they fall back to the bytes. Added `test_fetch_failure_is_terminal` covering timeout, connection, generic and status errors.

File: tests/frontend/test_http_status_propagation.py (hunk 2 above)
Comment (pskiran1) [reply tests/frontend/test_http_status_propagation.py]: Thank you. This was the CI failure too. The Rust dim probe refused the URL before the worker ran. I added a second `ModelInput.Text` registration so the request reaches `ImageLoader` with no preprocessing in front, and the test now asserts Python's message with the canary unchanged.

**Review by rmccorm4 — APPROVED** [review]

**Review by ishandhanani — APPROVED** [review]
